### ISSUE TITLE
offset and length option when reading object

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-stor#499]
To be used as minio backend, we need [client.Read](https://godoc.org/github.com/zero-os/0-stor/client#Client.Read) to have `start offset` and `length` parameter.

Amazon S3 API also allow this see `Range` header at https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html

**Comment by GlenDC** [comment]
That does seem like a high-level feature, not something we'll want to support on a datastor level.

We need to think about how we'll going to implement this, could be part of the regular client.Read, but would it also be fine @iwanbk to implement this feature as a new function called `client.ReadRange` instead?

**Comment by iwanbk** [comment]
to add more info, the `length` still not used when i downloaded 27M of file

**Comment by GlenDC** [comment]
> the length still not used when i downloaded 27M of file

What do you mean with that?

**Comment by iwanbk** [comment]
> What do you mean with that?

I thought minio client will download the big file piece by piece, but it turned out that the client download the whole 27M in single http request

**Comment by iwanbk** [comment]
> would it also be fine @iwanbk to implement this feature as a new function called client.ReadRange instead?

Yeap.

Although still not sure about the right implementation, i can see that `length` is possible.

But not sure about `offset`, i think 0-stor client still need to read from the start and only send data starting from the given `offset`, no?

**Comment by GlenDC** [comment]
Well, it depends. Both length and offset will be possible,
but in the current implementation it won't save anything optimal.
We'll still have to read/load most of the data, to then only return the selected part.

**Comment by lafont-e** [comment]
Badger does not seem to support the seek function, so or we emulate it as Glen suggested, ie: read the whole file, select the bytes from start to start+length, return that data and throw the rest of file, until the next request asks for another (start, length) piece of data, or we create a cache on the client that can handle such requests, the biggest problem is having 20 big files being asked for bits, the memory consumption will be unacceptable.

There is another option, look for an alternative to badger that already supports seek requests, or ask Badger for a FR, or create that function inside badger. 

The last option is to return a FAIL always that a seek is requested in the minio gateway, as this is not an important feature it can stay that way until needed.

**Comment by GlenDC** [comment]
> There is another option, look for an alternative to badger that already supports seek requests, or ask Badger for a FR, or create that function inside badger.

Could be a nice option, could you explore this? The badger team is also really accessible using Slack, but you can probably already figure a lot out reading the docs and source code.

With #514  in mind, does 0-db support a seek operation @maxux? Because if so, we could perhaps already provide this feature in our db interface, and simply make it fail for badger (for now).

> The last option is to return a FAIL always that a seek is requested in the minio gateway, as this is not an important feature it can stay that way until needed.

Can our minio integration feature work without problems, going for this route? If so, perhaps this should be our choice for the TFT launch, as we already have enough work as it is.

**Comment by lafont-e** [comment]
I'm looking for a public seek API function on the S3 bucket , and I'm finding none, so not sure seek is something that any user may ask over a file in a bucket. Yes, internally they use seek, not sure why but don't see in any of the gateways a reference to the Seeker interface or plain seek.

**Comment by iwanbk** [comment]
> I'm looking for a public seek API function on the S3 bucket , and I'm finding none,

In http request, `seek` is achieved through `range` header. See this issue description for more detail.

> Can our minio integration feature work without problems, going for this route?

FWIW seek is not used when downloading large (27 MB) file using these:
- minio client
- minio browser
- Go client  with default option

So i think we are OK with going for this route

**Comment by GlenDC** [comment]
Sorry @iwanbk, I wish it could be different, but I'll have to assign it you I'm afraid :( I'll be able to review though.

**Comment by GlenDC** [comment]
Closed by 7ab7260

**Comment by iwanbk** [comment]
Moved from https://github.com/zero-os/0-stor/issues/515#issuecomment-371134026

>> So, are you saying that if we have blockSize=4096 and we want to read from offset=5000, we can  simply fetch and decode from 2nd chunck?

> yes

It seems that this is not always the case.

I've uploaded 1.3G of file, this is what i have in meta

```
meta.Size = 1254203480
len(meta.Chunks) = 2916
```

It didn't use encryption & compression

`blockSize = 1048576`

i expected the number of chunks should be only 1254203480/1048576: 1196, no?

when i enabled compression and encryption, the number of chunks become 2922

**Comment by iwanbk** [comment]
This patch fixed my issue
```diff
+++ b/client/datastor/pipeline/async_splitter.go
@@ -813,7 +813,7 @@ func newAsyncDataSplitter(ctx context.Context, r io.Reader, chunkSize, bufferSiz
                var index int
                buf := make([]byte, chunkSize)
                for {
-                       n, err := r.Read(buf)
+                       n, err := io.ReadFull(r, buf)
                        if n > 0 {
                                data := make([]byte, n)
                                copy(data, buf)
@@ -825,7 +825,7 @@ func newAsyncDataSplitter(ctx context.Context, r io.Reader, chunkSize, bufferSiz
                                }
                        }
                        if err != nil {
-                               if err == io.EOF {
+                               if err == io.EOF || err == io.ErrUnexpectedEOF {
                                        // we'll consider an EOF
                                        // as a signal to let us know the reader is exhausted
                                        return nil
```

still need to do these:
- add some test
- check that treat `io.ErrUnexpectedEOF` the same as `io.EOF` is safe

**Comment by GlenDC** [comment]
> i expected the number of chunks should be only 1254203480/1048576: 1196, no?

1197*, as there is a bit too much data to fit in 1196 chunks as far as I can see.

> This patch fixed my issue

cool

**Comment by GlenDC** [comment]
fixed by PR #524.
